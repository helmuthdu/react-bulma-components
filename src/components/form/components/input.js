import cn from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';
import CONSTANTS from '../../../constants';
import modifiers from '../../../modifiers';
import { Element } from '../../element';

const colors = Object.values(CONSTANTS.COLORS);

export const Input = ({
  className,
  color,
  disabled,
  inactive,
  name,
  placeholder,
  readOnly,
  size,
  type,
  value,
  ...props
}) => (
  <Element
    {...props}
    renderAs="input"
    name={name}
    value={value}
    type={type}
    placeholder={placeholder}
    readOnly={readOnly || inactive}
    disabled={disabled}
    className={cn('input', className, {
      'is-static': inactive,
      [`is-${size}`]: size,
      [`is-${color}`]: color
    })}
  />
);

Input.propTypes = {
  ...modifiers.propTypes,
  className: PropTypes.string,
  color: PropTypes.oneOf(colors),
  disabled: PropTypes.bool,
  inactive: PropTypes.bool,
  name: PropTypes.string,
  placeholder: PropTypes.string,
  readOnly: PropTypes.bool,
  size: PropTypes.oneOf(Object.values(CONSTANTS.SIZES)),
  style: PropTypes.object,
  type: PropTypes.oneOf([
    'text',
    'email',
    'tel',
    'password',
    'number',
    'search',
    'color',
    'date',
    'time',
    'datetime-local'
  ]),
  value: PropTypes.string
};

Input.defaultProps = {
  ...modifiers.defaultProps,
  type: 'text',
  value: ''
};
