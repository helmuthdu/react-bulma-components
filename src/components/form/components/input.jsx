import cn from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';
import CONSTANTS from '../../../constants';
import modifiers from '../../../modifiers';
import { Element } from '../../element';

const colors = Object.values(CONSTANTS.COLORS);

export const Input = ({
  className,
  color,
  disabled,
  inactive,
  name,
  placeholder,
  readOnly,
  size,
  type,
  value,
  ...props
}) => (
  <Element
    {...props}
    disabled={disabled}
    name={name}
    placeholder={placeholder}
    readOnly={readOnly || inactive}
    renderAs="input"
    type={type}
    value={value}
    className={cn('input', className, {
      'is-static': inactive,
      [`is-${color}`]: color,
      [`is-${size}`]: size
    })}
  />
);

Input.propTypes = {
  ...modifiers.propTypes,
  className: PropTypes.string,
  color: PropTypes.oneOf(colors),
  disabled: PropTypes.bool,
  inactive: PropTypes.bool,
  name: PropTypes.string,
  placeholder: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  readOnly: PropTypes.bool,
  size: PropTypes.oneOf(Object.values(CONSTANTS.SIZES)),
  style: PropTypes.object,
  type: PropTypes.oneOf([
    'color',
    'date',
    'datetime-local',
    'email',
    'number',
    'password',
    'search',
    'tel',
    'text',
    'time'
  ]),
  value: PropTypes.string
};

Input.defaultProps = {
  ...modifiers.defaultProps,
  type: 'text',
  value: ''
};
