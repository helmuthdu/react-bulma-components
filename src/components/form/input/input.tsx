import cn from 'classnames';
import React from 'react';
import { Colors, Sizes } from '../../../constants';
import modifiers, { ElementModifier } from '../../../modifiers';
import { Element } from '../../elements';

type InputProps = Partial<Omit<React.ComponentProps<'input'>, 'color' | 'size' | 'unselectable'>> &
  ElementModifier & {
    color?: Colors;
    disabled?: boolean;
    inactive?: boolean;
    name?: string;
    placeholder?: string;
    onChange: (...args: any[]) => void;
    readOnly?: boolean;
    size?: Sizes;
    type?: 'text' | 'email' | 'tel' | 'password' | 'number' | 'search' | 'color' | 'date' | 'time' | 'datetime-local';
    value?: string | number;
  };

export const Input: React.FunctionComponent<InputProps> = ({
  className,
  color,
  disabled,
  inactive,
  name,
  placeholder,
  readOnly,
  size,
  type,
  value,
  ...props
}: InputProps) => (
  // @ts-ignore
  <Element
    {...props}
    disabled={disabled}
    name={name}
    placeholder={placeholder}
    readOnly={readOnly || inactive}
    renderAs="input"
    type={type}
    value={value}
    className={cn('input', className, {
      'is-static': inactive,
      [`is-${color}`]: color,
      [`is-${size}`]: size
    })}
  />
);

Input.defaultProps = {
  ...modifiers.defaultProps,
  type: 'text',
  value: ''
};
