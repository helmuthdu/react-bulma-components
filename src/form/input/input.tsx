import clsx from 'clsx';
import * as React from 'react';
import { ChangeEvent, useEffect, useState } from 'react';
import { Colors, Sizes } from '../../constants';
import { Element } from '../../elements';
import modifiers, { ElementModifier } from '../../modifiers';

type InputProps = ElementModifier & {
  color?: Colors;
  inactive?: boolean;
  readOnly?: boolean;
  rounded?: boolean;
  size?: Sizes;
  type?: 'text' | 'email' | 'tel' | 'password' | 'number' | 'search' | 'color' | 'date' | 'time' | 'datetime-local';
  value?: string | number;
} & Omit<React.ComponentProps<'input'>, 'ref' | 'color' | 'size' | 'unselectable' | 'value'>;

export const Input = React.forwardRef<HTMLInputElement, InputProps>(
  (
    {
      className,
      color,
      disabled,
      inactive,
      name,
      placeholder,
      readOnly,
      rounded,
      size,
      type,
      value: initialValue = '',
      onChange,
      ...props
    },
    ref
  ) => {
    const [value, setValue] = useState(initialValue);

    useEffect(() => {
      setValue(initialValue);
    }, [initialValue]);

    return (
      <Element
        ref={ref}
        disabled={disabled}
        name={name}
        placeholder={placeholder}
        readOnly={readOnly || inactive}
        as="input"
        type={type}
        value={value}
        onChange={(evt: ChangeEvent<HTMLInputElement>) => {
          setValue(evt.target.value);
          if (onChange) onChange(evt);
        }}
        className={clsx('input', className, {
          'is-static': inactive,
          'is-rounded': rounded,
          [`is-${color}`]: color,
          [`is-${size}`]: size
        })}
        {...props}
      />
    );
  }
);

Input.defaultProps = {
  ...modifiers.defaultProps,
  type: 'text',
  value: ''
};
