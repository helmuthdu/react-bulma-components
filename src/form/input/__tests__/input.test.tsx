import { fireEvent, render } from '@testing-library/react';
import * as React from 'react';
import { Input } from '../input';

describe('Input component', () => {
  let onChange: jest.Mock;

  beforeEach(() => {
    onChange = jest.fn();
  });

  it('should render', () => {
    const { asFragment } = render(<Input onChange={onChange} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should concat classname in props with classname', () => {
    const { asFragment } = render(<Input onChange={onChange} className="other-class this-is-a-test" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should use inline styles', () => {
    const { asFragment } = render(<Input onChange={onChange} style={{ height: 250 }} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should be type email and a with success colors', () => {
    const { container } = render(<Input onChange={onChange} color="success" type="email" />);
    const input = container.querySelector('input') as HTMLInputElement;
    expect(input).toHaveClass('is-success');
    expect(input.type).toEqual('email');
  });

  it('should be rounded, large, inactive and readOnly', () => {
    const { container } = render(<Input onChange={onChange} readOnly size="large" rounded inactive />);
    const input = container.querySelector('input') as HTMLInputElement;
    expect(input).toHaveClass('is-large', 'is-rounded', 'is-static');
    expect(input.readOnly).toEqual(true);
  });

  it('should be disabled with placeholder and value', () => {
    const { container } = render(<Input onChange={onChange} value="TEST" disabled placeholder="hello tests" />);
    const input = container.querySelector('input') as HTMLInputElement;
    expect(input.value).toEqual('TEST');
    expect(input.disabled).toEqual(true);
    expect(input.placeholder).toEqual('hello tests');
  });

  it('should update value', () => {
    const { container } = render(<Input onChange={onChange} value="TEST" disabled placeholder="hello tests" />);
    const input = container.querySelector('input') as HTMLInputElement;
    fireEvent.change(input, { target: { value: 'new value' } });
    expect(input).toHaveValue('new value');
  });
});
