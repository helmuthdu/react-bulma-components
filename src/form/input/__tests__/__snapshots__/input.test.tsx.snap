// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component should concat classname in props with classname 1`] = `
<DocumentFragment>
  <input
    class="input other-class this-is-a-test"
    type="text"
    value=""
  />
</DocumentFragment>
`;

exports[`Input component should render 1`] = `
<DocumentFragment>
  <input
    class="input"
    type="text"
    value=""
  />
</DocumentFragment>
`;

exports[`Input component should use inline styles 1`] = `
<DocumentFragment>
  <input
    class="input"
    style="height: 250px;"
    type="text"
    value=""
  />
</DocumentFragment>
`;
